Added KUKA IIWA Robot model, dynamics simulation, and unit test

liangfok opened the pull request:
Closes https://github.com/RobotLocomotion/drake/issues/2516 and https://github.com/RobotLocomotion/drake/pull/2163.

Here's what the beautiful robot looks like after it collapses on the ground:

<img width="595" alt="screen shot 2016-06-08 at 10 34 22 am" src="https://cloud.githubusercontent.com/assets/1388098/15898019/a30a0fec-2d64-11e6-9e2a-4125b03e9aee.png">


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2520)
<!-- Reviewable:end -->

liangfok commented:
+@jwnimmer-tri for both levels of review since this is a relatively simple PR that doesn't touch any existing code.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> #### Added KUKA IIWA Robot model, dynamics simulation, and unit test
> 
> Closes https://github.com/RobotLocomotion/drake/issues/2516 and https://github.com/RobotLocomotion/drake/pull/2163.
> 
> 
> 
</details>


---

Review status: 0 of 21 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-:-KJkQ8sk-pSffdamj2-9:1745428341)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@sammy-tri for feature review because I recall he was interested in having this for his research into solvers.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> +@jwnimmer-tri for both levels of review since this is a relatively simple PR that doesn't touch any existing code.
</details>


---

Review status: 0 of 21 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-:-KJkVAgQvqrIpZ-vPSh_:-1286297749)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 36 of 41 files at r2.
Review status: 18 of 21 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/urdf/iiwa14.urdf, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkiIKMRZLgRJJFjkIl:-KJkiIKMRZLgRJJFjkIm:-669686496) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/iiwa14.urdf#L1)):*
> ```XML
> <?xml version="1.0" ?>
> ```

BTW I did not review the content of this file.

---

*[drake/examples/kuka_iiwa_arm/urdf/info.txt, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkehj2T3Psv78DNjBy:-KJkehj3IQIUjecvuhAe:-809855472) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/info.txt#L1)):*
> ```
> Kuka LBR IIWA R820 14. Arm model taken from: https://github.com/SalvoVirga/iiwa_stack
> ```

Missing newline at end of file.

---

*[drake/examples/kuka_iiwa_arm/urdf/info.txt, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkejlbvklzm-O5np2s:-KJkejlbvklzm-O5np2t:110057436) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/info.txt#L1)):*
> ```
> Kuka LBR IIWA R820 14. Arm model taken from: https://github.com/SalvoVirga/iiwa_stack
> ```

Are the obj and urdf files licensed under the same terms as Drake, and owned by the same authors as Drake?  If not, the alternative license and ownership needs to be documented in this directory.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 17 of 22 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/urdf/info.txt, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkehj2T3Psv78DNjBy:-KJkkRYLd3HPGrOfsYf8:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/info.txt#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing newline at end of file.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/urdf/info.txt, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkejlbvklzm-O5np2s:-KJkkTsI-O3yddZvZjTD:1662589556) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/info.txt#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Are the obj and urdf files licensed under the same terms as Drake, and owned by the same authors as Drake?  If not, the alternative license and ownership needs to be documented in this directory.
</details>
I added a LICENSE.TXT. Hope I did it right. Do we need to consult a software IP lawyer for this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot test this please

jamiesnape commented:
@drake-jenkins-bot test this please

jwnimmer-tri commented:
Reviewed 2 of 41 files at r2, 1 of 1 files at r3, 2 of 2 files at r4.
Review status: 21 of 22 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 8 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkjas3UqtWkJRxMnas:-KJkjas3UqtWkJRxMnat:-911671502) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L8)):*
> ```objc
> 
> /// Implements a Drake System (@see drake/systems/System.h) that
> /// saves the latest system state as a local variable and provides an accessor
> ```

The use  of `state` here is confusing.  The tap saves the input, not the state (there is no state).

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 14 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkikPMxxWKGTrIS5OI:-KJkikPMxxWKGTrIS5OJ:979860490) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L14)):*
> ```objc
> class RobotStateTap {
>  public:
>   /// Create an RobotStateTap that publishes on the given @p lcm instance.
> ```

Comment about `lcm`is a lie.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 15 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkiqUJSGCLAFjj3n8_:-KJkiqUKvoT-R2I49YYh:470973925) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L15)):*
> ```objc
>  public:
>   /// Create an RobotStateTap that publishes on the given @p lcm instance.
>   explicit RobotStateTap() {}
> ```

`explicit` is unwarranted here (no args).

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 39 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkkJ-iaD5Wm_IXuD9W:-KJkkJ-jyljhNtDJR7nd:1544914228) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L39)):*
> ```objc
>                               const InputVector<double>& u) {
>     if (u_.size() != u.size())
>       u_.resize(u.size());
> ```

The `resize` method is not part of the Vector concept.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 42 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkkNr9CbPrWytmpVwy:-KJkkNr9CbPrWytmpVwz:260001251) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L42)):*
> ```objc
> 
>     for (int ii = 0; ii < u.size(); ++ii) {
>       u_[ii] = u[ii];
> ```

The `operator[]` method is not part of the Vector concept (though maybe it should be?).

In any case, using `operator=(Eigen)` and `toEigen`  would likely the best implementation choice here anyway, instead of resize + loop.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 51 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkjRu4Odeu8OdZXebE:-KJkjRu4Odeu8OdZXebF:-436511042) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L51)):*
> ```objc
>   bool isDirectFeedthrough() const { return true; }
> 
>   const InputVector<double>& get_input_vector() { return u_; }
> ```

This is within the "Implement the Drake System" comment block, but in fact is not part of the System concept.

---

*[drake/examples/kuka_iiwa_arm/urdf/info.txt, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkejlbvklzm-O5np2s:-KJl4lzGiAP3gp7MlzQD:797934947) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/info.txt#L1)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I added a LICENSE.TXT. Hope I did it right. Do we need to consult a software IP lawyer for this?
</details>
It looks good enough to me.  The 3-clause BSD they used it a bit sad, but not a problem.

---

*[drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT, line 3 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl2MI8BYbqatTr2jco:-KJl2MI925bTJxtSKckw:-843688789) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT#L3)):*
> ```
> The URDF and OBJ files were taken and then modified from:
> https://github.com/SalvoVirga/iiwa_stack/
> The original license for these files is given below. Our changes are covered by Drake's license terms (See LICENSE.TXt in Drake's root folder).
> ```

Typo `TXt`.
Long line.

---

*[drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT, line 36 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl3O3uuVIhI_BlSP-z:-KJl3O3uuVIhI_BlSP0-:449016398) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT#L36)):*
> ```
> 
> Most of the files were drastically modified, but the original authors inspired
> our work. For everything else :
> ```

This paragraph is confusing.  I see now that it's upstream's "our work" means salvo, not drake developers.  Suggestion: put `iiwa_stack.LICENSE.txt` into Drake unmodified, and then have Drake's LICENSE.txt in this directory just refer to it by filename.  Mixing our caveat text atop this file into their caveat text from their license file is unclear.

I think even the info.txt content could move into our `./LICENSE.txt` here to have fewer files.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot test this please

jwnimmer-tri commented:
Reviewed 1 of 41 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl6WoyzHjRufwOEZhp:-KJl6WoyzHjRufwOEZhq:1821586227) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L4)):*
> ```CMake
>   add_executable(run_kuka_iiwa_arm_dynamics run_kuka_iiwa_arm_dynamics.cc)
>   target_link_libraries(run_kuka_iiwa_arm_dynamics drakeRBSystem drakeLCMSystem)
>   add_test(NAME run_kuka_iiwa_arm_dynamics COMMAND run_kuka_iiwa_arm_dynamics --non-realtime)
> ```

For reference -- approximately how long does the demo (test) take to run on your computer?

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl8LLGPOy0vWH-pDV-:-KJl8LLGPOy0vWH-pDV0:1832810651) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L4)):*
> ```CMake
>   add_executable(run_kuka_iiwa_arm_dynamics run_kuka_iiwa_arm_dynamics.cc)
>   target_link_libraries(run_kuka_iiwa_arm_dynamics drakeRBSystem drakeLCMSystem)
>   add_test(NAME run_kuka_iiwa_arm_dynamics COMMAND run_kuka_iiwa_arm_dynamics --non-realtime)
> ```

I don't see how or where `--non-realtime` is being used by the test program?

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 31 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl5KY2Fnt2GXyqn1Ig:-KJl5KY3BTqjL5kVg7pE:-2111354431) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L31)):*
> ```cpp
> 
>   // Obtains a reference to the rigid body tree within the rigid body system.
>   auto const& tree = rigid_body_sys->getRigidBodyTree();
> ```

`const auto&`

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 79 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl64PDjGJLHal1ltOw:-KJl64PDjGJLHal1ltOx:1134367053) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L79)):*
> ```cpp
> 
>   // Prevents exception from being thrown when simulation runs slower than real
>   // time, which it most likely will given the small step size.
> ```

In which case, why not have realtime_factor "as fast as possible"?  I guess you don't want it to ever run faster than realtime?

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 148 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl7aiqlS1HiahUZNfo:-KJl7airHxOBRl2wcmWr:837191321) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L148)):*
> ```cpp
>                                  std::to_string(position) + " > " +
>                                  std::to_string(max_limit[ii]) + ").");
>       }
> ```

BTW I guess this kind of "test" via a demo example is fine insofar as it goes, but it would be nice if in the future we can find a way to use googletest for these kind of comparison statements.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -- marking as LGTM because I'm not going to be available for the next few days and I think @jwnimmer-tri's existing comments capture any concerns which I had while reading this PR.  If they all get resolved to his satisfaction, I'm happy.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> +@sammy-tri for feature review because I recall he was interested in having this for his research into solvers.
</details>


---

Reviewed 3 of 41 files at r2, 1 of 1 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/urdf/iiwa14.urdf, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkiIKMRZLgRJJFjkIl:-KJlOWOXrjL99OgQZQ0z:-1971990920) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/iiwa14.urdf#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I did not review the content of this file.
</details>
Neither did I...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-:-KJlOliP89QuxEiBse_Q:259804188)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL.

<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> :lgtm: -- marking as LGTM because I'm not going to be available for the next few days and I think @jwnimmer-tri's existing comments capture any concerns which I had while reading this PR.  If they all get resolved to his satisfaction, I'm happy.
</details>


---

Review status: 17 of 22 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl8LLGPOy0vWH-pDV-:-KJlFPSQXjDjum3orSyh:-1352561846) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I don't see how or where `--non-realtime` is being used by the test program?
</details>
Oops. I forgot to remove that. Removed.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl6WoyzHjRufwOEZhp:-KJlG4JuQvnrZH8xGw3s:278893854) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> For reference -- approximately how long does the demo (test) take to run on your computer?
</details>
Using the `time` command, here are two data points:

```
Total Test time (real) =  66.93 sec

real	1m6.931s
user	1m6.840s
sys	0m0.152s

Total Test time (real) =  69.61 sec

real	1m9.614s
user	1m8.760s
sys	0m0.884s
```

Is this too long? I can always decrease the duration of the experiment.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 8 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkjas3UqtWkJRxMnas:-KJlGViMaJxcIj3pAMIz:-125746836) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L8)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The use  of `state` here is confusing.  The tap saves the input, not the state (there is no state).
</details>
Done. Modified it to state that it saves the "input" instead of the "system state".

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 14 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkikPMxxWKGTrIS5OI:-KJlGZk7kZOOVomCVT0X:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L14)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Comment about `lcm`is a lie.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 15 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkiqUJSGCLAFjj3n8_:-KJlGcBUjygJOVhURW0O:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> `explicit` is unwarranted here (no args).
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 39 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkkJ-iaD5Wm_IXuD9W:-KJlObBpd6FsvodDpdHq:878446086) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L39)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The `resize` method is not part of the Vector concept.
</details>
Done. Replaced with `Drake::toEigen(u)` as recommended.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 42 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkkNr9CbPrWytmpVwy:-KJlRpHLAL82wbRz9GiY:-352158138) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L42)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The `operator[]` method is not part of the Vector concept (though maybe it should be?).
> 
> In any case, using `operator=(Eigen)` and `toEigen`  would likely the best implementation choice here anyway, instead of resize + loop.
</details>
Done. Replaced with `Drake::toEigen(u)` as recommended.

---

*[drake/examples/kuka_iiwa_arm/robot_state_tap.h, line 51 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkjRu4Odeu8OdZXebE:-KJlRr8fBlpWFk_sOWc5:-1131897488) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/robot_state_tap.h#L51)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This is within the "Implement the Drake System" comment block, but in fact is not part of the System concept.
</details>
Done. Moved it outside of the comment block.

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 31 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl5KY2Fnt2GXyqn1Ig:-KJlRwZjGFiuBW8YRo0D:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L31)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> `const auto&`
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 79 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl64PDjGJLHal1ltOw:-KJlS1N_lfZZbTWQR747:-633059116) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L79)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> In which case, why not have realtime_factor "as fast as possible"?  I guess you don't want it to ever run faster than realtime?
</details>
I changed the real time factor to be "as fast as possible".

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 148 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl7aiqlS1HiahUZNfo:-KJlSE2YuA7mAIme9wB0:444560066) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L148)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I guess this kind of "test" via a demo example is fine insofar as it goes, but it would be nice if in the future we can find a way to use googletest for these kind of comparison statements.
</details>
Yeah, I agree. I can factor out common code between the example application and a corresponding Google Test and create two executables. Should I do that in this PR?

---

*[drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT, line 3 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl2MI8BYbqatTr2jco:-KJlSy2sjBxUt0tmoLb7:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT#L3)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Typo `TXt`.
> Long line.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT, line 36 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl3O3uuVIhI_BlSP-z:-KJlTmXgT4B6pcqV9RNc:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/urdf/LICENSE.TXT#L36)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This paragraph is confusing.  I see now that it's upstream's "our work" means salvo, not drake developers.  Suggestion: put `iiwa_stack.LICENSE.txt` into Drake unmodified, and then have Drake's LICENSE.txt in this directory just refer to it by filename.  Mixing our caveat text atop this file into their caveat text from their license file is unclear.
> 
> I think even the info.txt content could move into our `./LICENSE.txt` here to have fewer files.
</details>
Done.

---

*[drake/examples/kuka_iiwa_arm/urdf/info.txt, line 1 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJkejlbvklzm-O5np2s:-KJlToVfnLk6SDV3PAKk:-528723737) ([raw file](https://github.com/robotlocomotion/drake/blob/c9812fca354a93123a184ab96abd0aec84b90145/drake/examples/kuka_iiwa_arm/urdf/info.txt#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> It looks good enough to me.  The 3-clause BSD they used it a bit sad, but not a problem.
</details>
Yeah, I guess it's too much trouble trying to get the original authors to issue another license.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-:-KJlVBV8NC_1NocgXtor:-84625242)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 41 files at r2, 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl6WoyzHjRufwOEZhp:-KJld640TU-ILb627P_7:362640229) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L4)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Using the `time` command, here are two data points:
> 
> ```
> Total Test time (real) =  66.93 sec
> 
> real	1m6.931s
> user	1m6.840s
> sys	0m0.152s
> 
> Total Test time (real) =  69.61 sec
> 
> real	1m9.614s
> user	1m8.760s
> sys	0m0.884s
> ```
> 
> Is this too long? I can always decrease the duration of the experiment.
</details>
Yeah, that's pretty long.  Maybe have ctest pass in a short duration (0.5 second realtime = 6 second cpu time)?  Is that enough to cover everything well enough?

The default empty (demo) cmdline can still use 5 seconds if you like.

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 148 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl7aiqlS1HiahUZNfo:-KJldZ9aw5sgNlr5a2Vi:240812607) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L148)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Yeah, I agree. I can factor out common code between the example application and a corresponding Google Test and create two executables. Should I do that in this PR?
</details>
I think merging this one, without any large refactoring, is my preference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed all reviewer comments. PTAL.
</details>


---

Review status: 20 of 22 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 4 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl6WoyzHjRufwOEZhp:-KJlfjzCL0rFGOLrwvjS:906114880) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Yeah, that's pretty long.  Maybe have ctest pass in a short duration (0.5 second realtime = 6 second cpu time)?  Is that enough to cover everything well enough?
> 
> The default empty (demo) cmdline can still use 5 seconds if you like.
</details>
Done. I modified the unit test to run for 0.5 seconds. Here are the new benchmark results:

```
$ time ctest -VV -R run_kuka_iiwa_arm_dynamics 
...
Total Test time (real) =   6.61 sec

real	0m6.612s
user	0m6.528s
sys	0m0.088s

$ time ./pod-build/bin/run_kuka_iiwa_arm_dynamics
...
real	1m2.241s
user	1m2.172s
sys	0m0.116s

```

---

*[drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc, line 148 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-KJl7aiqlS1HiahUZNfo:-KJlhHaBkn_PgEzcMDjc:-486510564) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1903d566ff0bf8ff92fb7452e5fe60cfb2bd53/drake/examples/kuka_iiwa_arm/run_kuka_iiwa_arm_dynamics.cc#L148)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think merging this one, without any large refactoring, is my preference.
</details>
SGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-:-KJliQFECiFjf2H3VzgS:-84625242)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/1978/ and is thus ready for final review / merge.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I addressed all reviewer comments. PTAL.
</details>


---

Review status: 20 of 22 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-:-KJlwsKFfysXoPeK8Mbv:750525280)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/1978/ and is thus ready for final review / merge.
</details>


---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2520#-:-KJlz6YhtRj9NMpJUg3b:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Using override-merge due to out-of-date-with-master.
